### ISSUE TITLE
`experimental.odeint` custom gradient rule breaks with integer arguments

### ISSUE THREAD

**Issue by PhilipVinc** [google/jax#16517]
### Description

I know that `experimental.odeint` is in a sense discontinued, but maybe there is an easy fix for this?

MWE:
```python
from jax.experimental.ode import odeint
import jax
import jax.numpy as jnp

def f(y, t, arg1, arg2):
    return -y*arg1 + arg2

y0 = jnp.array([2., 3.])
ts = jnp.array([0.0, 1.0])
arg1 = jnp.array(0.2)
arg2 = jnp.array(1, dtype=int)
sol = odeint(f, y0, ts, arg1, arg2)

def get_sol(arg1):
    # note that arg2 is captured so we don't differentiate wrt it.
    sol = odeint(f, y0, ts, arg1, arg2)
    return sol

jax.jacobian(get_sol)(arg1)
```

This happens because odeint's custom gradient rule attempts to compute the gradient wrt all arguments, even `arg2` which is an integer and I was not trying to actually differentiate.

While this might seem nonsensical, this case shows up if you try to integrate with odeint a differential equation defined by a sparse matrix, which has the structure encoded in integers.

### What jax/jaxlib version are you using?

jax 0.4.11
